File: implementations/go/s57_test.go
package b57

import (
	"bytes"
	"errors"
	"testing"
)

func testS57(t *testing.T) *S57 {
	t.Helper()
	s, err := NewS57(S57Config{
		ServerSecretKey: []byte("S57_SERVER_SECRET_KEY_MUST_BE_LONG_1234567890"),
		EnvironmentSalt: []byte("prod-v1"),
		KeyID:           7,
	})
	if err != nil {
		t.Fatalf("NewS57 failed: %v", err)
	}
	return s
}

func TestS57ResolveKeys(t *testing.T) {
	s := testS57(t)
	keys, err := s.ResolveKeys(7)
	if err != nil {
		t.Fatalf("ResolveKeys failed: %v", err)
	}
	if len(keys.B57AES256Key) != 32 || len(keys.H57Key) != 32 || len(keys.ID57Key) != 32 {
		t.Fatalf("invalid key lengths")
	}
}

func TestS57HashDeterministic(t *testing.T) {
	s := testS57(t)
	in := []byte("hello-s57-hash")
	a, err := s.Hash(in, H57Len256)
	if err != nil {
		t.Fatalf("Hash failed: %v", err)
	}
	b, err := s.Hash(in, H57Len256)
	if err != nil {
		t.Fatalf("Hash failed: %v", err)
	}
	if a != b {
		t.Fatalf("hash should be deterministic")
	}
	if len(a) != 44 {
		t.Fatalf("expected 44 chars, got %d", len(a))
	}
}

func TestS57IDDeterministicAndLengths(t *testing.T) {
	s := testS57(t)
	in := []byte("hello-s57-id")
	v128, err := s.ID(in, ID57Default)
	if err != nil {
		t.Fatalf("ID failed: %v", err)
	}
	v256, err := s.ID(in, ID57Len256)
	if err != nil {
		t.Fatalf("ID failed: %v", err)
	}
	v512, err := s.ID(in, ID57Len512)
	if err != nil {
		t.Fatalf("ID failed: %v", err)
	}
	min128, max128, err := ID57Range(ID57Len128)
	if err != nil {
		t.Fatalf("ID57Range(ID57Len128) failed: %v", err)
	}
	min256, max256, err := ID57Range(ID57Len256)
	if err != nil {
		t.Fatalf("ID57Range(ID57Len256) failed: %v", err)
	}
	min512, max512, err := ID57Range(ID57Len512)
	if err != nil {
		t.Fatalf("ID57Range(ID57Len512) failed: %v", err)
	}

	if len(v128) < min128 || len(v128) > max128 {
		t.Fatalf("v128 length %d outside bound [%d,%d]", len(v128), min128, max128)
	}
	if len(v256) < min256 || len(v256) > max256 {
		t.Fatalf("v256 length %d outside bound [%d,%d]", len(v256), min256, max256)
	}
	if len(v512) < min512 || len(v512) > max512 {
		t.Fatalf("v512 length %d outside bound [%d,%d]", len(v512), min512, max512)
	}

	// S57 restricts length_enum to the 128/256/512-bit security profile;
	// any other positive bit length (and, by extension, any fixed width -
	// non-security by design) MUST be rejected.
	if _, err := s.ID(in, ID57Length(47)); err == nil {
		t.Fatalf("expected invalid length error for S57 profile")
	}
	if _, err := s.ID(in, ID57Fixed8); err == nil {
		t.Fatalf("expected S57 to reject fixed (non-security) lengths")
	}
}

func TestS57RandomAPIs(t *testing.T) {
	s := testS57(t)
	values := []string{}
	add := func(v string, err error) {
		if err != nil {
			t.Fatalf("random api failed: %v", err)
		}
		values = append(values, v)
	}
	add(s.Random())
	add(s.RandomTime())
	add(s.RandomCounter(42))
	add(s.RandomSession([]byte("session-secret")))
	add(s.RandomDevice([]byte("device-secret")))
	add(s.RandomDerived([]byte("master-secret"), []byte("u-1")))
	add(s.RandomHardened())
	add(s.RandomHybrid([]byte("a"), []byte("b"), []byte("c")))

	for i, v := range values {
		if len(v) != 22 {
			t.Fatalf("value %d expected len 22, got %d", i, len(v))
		}
		if !IsValid(v) || !IsCanonical(v) {
			t.Fatalf("value %d is not valid canonical B57", i)
		}
	}
}

func TestS57RandomDerivedDeterministic(t *testing.T) {
	s := testS57(t)
	a, err := s.RandomDerived([]byte("master"), []byte("unique"))
	if err != nil {
		t.Fatalf("RandomDerived failed: %v", err)
	}
	b, err := s.RandomDerived([]byte("master"), []byte("unique"))
	if err != nil {
		t.Fatalf("RandomDerived failed: %v", err)
	}
	if a != b {
		t.Fatalf("RandomDerived must be deterministic")
	}
}

func TestS57EncryptDecryptRoundTrip(t *testing.T) {
	s := testS57(t)
	pt := []byte("secret payload")
	aad := []byte("aad-v1")

	enc, err := s.Encrypt(pt, aad)
	if err != nil {
		t.Fatalf("Encrypt failed: %v", err)
	}
	out, err := s.Decrypt(enc, aad)
	if err != nil {
		t.Fatalf("Decrypt failed: %v", err)
	}
	if !bytes.Equal(out, pt) {
		t.Fatalf("decrypt output mismatch")
	}
}

func TestS57DecryptWrongAADFails(t *testing.T) {
	s := testS57(t)
	enc, err := s.Encrypt([]byte("hello"), []byte("good-aad"))
	if err != nil {
		t.Fatalf("Encrypt failed: %v", err)
	}
	if _, err := s.Decrypt(enc, []byte("bad-aad")); err == nil {
		t.Fatalf("expected auth failure")
	}
}

func TestS57DecryptWrongKeyIDFails(t *testing.T) {
	s := testS57(t)
	enc, err := s.Encrypt([]byte("hello"), nil)
	if err != nil {
		t.Fatalf("Encrypt failed: %v", err)
	}
	if _, err := s.Decrypt(enc, nil, 1); err == nil {
		t.Fatalf("expected key unavailable error")
	}
}

func TestS57DecryptInvalidVersionFails(t *testing.T) {
	s := testS57(t)
	enc, err := s.Encrypt([]byte("hello"), nil)
	if err != nil {
		t.Fatalf("Encrypt failed: %v", err)
	}
	raw, err := Decode(enc)
	if err != nil {
		t.Fatalf("Decode failed: %v", err)
	}
	raw[0] = 0xff
	tampered := Encode(raw)
	if _, err := s.Decrypt(tampered, nil); err == nil {
		t.Fatalf("expected invalid version error")
	}
}

func TestS57NewInvalidConfigFails(t *testing.T) {
	if _, err := NewS57(S57Config{ServerSecretKey: []byte("short"), EnvironmentSalt: []byte("x")}); err == nil {
		t.Fatalf("expected key invalid error for short server secret")
	}
	if _, err := NewS57(S57Config{ServerSecretKey: []byte("S57_SERVER_SECRET_KEY_MUST_BE_LONG_1234567890"), EnvironmentSalt: nil}); err == nil {
		t.Fatalf("expected key invalid error for missing salt")
	}
}

func TestS57ResolveKeysInvalidSeedFails(t *testing.T) {
	s := &S57{}
	if _, err := s.ResolveKeys(1); err == nil {
		t.Fatalf("expected error for invalid seed")
	}
}

func TestS57HashInvalidLengthFails(t *testing.T) {
	s := testS57(t)
	if _, err := s.Hash([]byte("x"), H57Len47); err == nil {
		t.Fatalf("expected invalid length enum error")
	}
}

func TestS57EncryptWithInvalidKeyFails(t *testing.T) {
	s := testS57(t)
	s.keys.B57AES256Key = nil
	if _, err := s.Encrypt([]byte("hello"), nil); err == nil {
		t.Fatalf("expected invalid key error")
	}
}

func TestS57DecryptInvalidEnvelopeLengthFails(t *testing.T) {
	s := testS57(t)
	if _, err := s.Decrypt("A", nil); err == nil {
		t.Fatalf("expected invalid envelope error")
	}
}

func TestS57DecryptWithInvalidStoredKeyFails(t *testing.T) {
	s := testS57(t)
	enc, err := s.Encrypt([]byte("hello"), nil)
	if err != nil {
		t.Fatalf("Encrypt failed: %v", err)
	}
	s.keys.B57AES256Key = nil
	if _, err := s.Decrypt(enc, nil); err == nil {
		t.Fatalf("expected key unavailable error")
	}
}

func TestS57EntropyFailures(t *testing.T) {
	s := testS57(t)
	original := s57EntropyReader
	s57EntropyReader = func(length int) ([]byte, error) {
		return nil, errors.New("entropy source failure")
	}
	defer func() { s57EntropyReader = original }()

	if _, err := s.Random(); err == nil {
		t.Fatalf("expected random entropy failure")
	}
	if _, err := s.RandomTime(); err == nil {
		t.Fatalf("expected random_time entropy failure")
	}
	if _, err := s.RandomCounter(1); err == nil {
		t.Fatalf("expected random_counter entropy failure")
	}
	if _, err := s.RandomSession([]byte("x")); err == nil {
		t.Fatalf("expected random_session entropy failure")
	}
	if _, err := s.RandomDevice([]byte("x")); err == nil {
		t.Fatalf("expected random_device entropy failure")
	}
	if _, err := s.RandomHardened(); err == nil {
		t.Fatalf("expected random_hardened entropy failure")
	}
	if _, err := s.Encrypt([]byte("x"), nil); err == nil {
		t.Fatalf("expected encrypt entropy failure")
	}
}

File: implementations/javascript/s57.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import { encode, decode, isCanonical, isValid } from './b57.js';
import { H57Length } from './h57.js';
import { ID57Length } from './id57.js';
import { ErrorCode } from './errors.js';
import { S57, S57_VERSION } from './s57.js';

const enc = new TextEncoder();

function createS57() {
  return new S57({
    server_secret_key: enc.encode('S57_SERVER_SECRET_KEY_MUST_BE_LONG_1234567890'),
    environment_salt: enc.encode('prod-v1'),
    key_id: 7
  });
}

test('S57 key derivation outputs 32-byte keys', () => {
  const s57 = createS57();
  const keys = s57.resolveKeys();
  assert.equal(keys.b57_aes_256_key.length, 32);
  assert.equal(keys.h57_key.length, 32);
  assert.equal(keys.id57_key.length, 32);
  assert.equal(keys.key_id, 7);
});

test('S57 hash deterministic and canonical', () => {
  const s57 = createS57();
  const input = enc.encode('hello-s57-hash');
  const a = s57.hash(input, H57Length.LEN_256);
  const b = s57.hash(input, H57Length.LEN_256);

  assert.equal(a, b);
  assert.equal(a.length, 44);
  assert.equal(isValid(a), true);
  assert.equal(isCanonical(a), true);
});

test('S57 id deterministic and enforces profile lengths', () => {
  const s57 = createS57();
  const input = enc.encode('hello-s57-id');

  const d = s57.id(input, ID57Length.DEFAULT);
  const v256 = s57.id(input, ID57Length.LEN_256);
  const v512 = s57.id(input, ID57Length.LEN_512);

  assert.equal(d.length, 22);
  assert.equal(v256.length, 44);
  assert.equal(v512.length, 88);

  // S57 only exposes the 128/256/512-bit security profile: non-security
  // fixed widths (FIXED_k, negative length_enum) MUST be rejected, since
  // they are non-security by definition (spec s57-security-57.txt 8).
  assert.throws(() => s57.id(input, ID57Length.FIXED_8), (err) => err.code === ErrorCode.INVALID_LENGTH_ENUM);
  assert.throws(() => s57.id(input, 77777), (err) => err.code === ErrorCode.INVALID_LENGTH_ENUM);
});

test('S57 random API variants produce canonical 22-char strings', () => {
  const s57 = createS57();
  const values = [
    s57.random(),
    s57.random_time(),
    s57.random_counter(42),
    s57.random_session(enc.encode('session-secret')),
    s57.random_device(enc.encode('device-secret')),
    s57.random_derived(enc.encode('master-secret'), enc.encode('u-1')),
    s57.random_hardened(),
    s57.random_hybrid(enc.encode('a'), enc.encode('b'), enc.encode('c'))
  ];

  for (const value of values) {
    assert.equal(value.length, 22);
    assert.equal(isValid(value), true);
    assert.equal(isCanonical(value), true);
  }
});

test('S57 random_derived is deterministic for same inputs', () => {
  const s57 = createS57();
  const master = enc.encode('master');
  const unique = enc.encode('unique');

  const a = s57.random_derived(master, unique);
  const b = s57.random_derived(master, unique);
  assert.equal(a, b);
});

test('S57 encrypt/decrypt round-trip with aad', () => {
  const s57 = createS57();
  const plaintext = enc.encode('secret payload');
  const aad = enc.encode('aad-v1');

  const cipher = s57.encrypt(plaintext, aad);
  const out = s57.decrypt(cipher, aad);

  assert.deepEqual(Array.from(out), Array.from(plaintext));
});

test('S57 decrypt fails with wrong aad', () => {
  const s57 = createS57();
  const plaintext = enc.encode('hello');
  const cipher = s57.encrypt(plaintext, enc.encode('good-aad'));

  assert.throws(
    () => s57.decrypt(cipher, enc.encode('bad-aad')),
    (err) => err.code === ErrorCode.AUTH_FAILURE
  );
});

test('S57 decrypt fails for unknown key_id', () => {
  const s57 = createS57();
  const plaintext = enc.encode('hello');
  const cipher = s57.encrypt(plaintext);

  assert.throws(
    () => s57.decrypt(cipher, new Uint8Array(0), 1),
    (err) => err.code === ErrorCode.KEY_UNAVAILABLE
  );
});

test('S57 decrypt fails for invalid version byte', () => {
  const s57 = createS57();
  const plaintext = enc.encode('hello');
  const cipher = s57.encrypt(plaintext);
  const raw = Uint8Array.from(decode(cipher));
  raw[0] = 0xff;
  const tampered = encode(raw);

  assert.throws(
    () => s57.decrypt(tampered),
    (err) => err.code === ErrorCode.INVALID_VERSION
  );
});

test('S57 constants exposed', () => {
  assert.equal(S57_VERSION, 0x01);
});
